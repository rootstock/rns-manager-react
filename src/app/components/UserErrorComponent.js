import React from 'react';
import propTypes from 'prop-types';
import { multilanguage } from 'redux-multilanguage';

import { ERROR_SAME_VALUE } from '../tabs/newAdmin/types';
import { ERROR_RESOLVE_NAME } from '../tabs/resolve/types';
import { ADDRESS_ENCODING_ERROR } from '../tabs/newAdmin/addresses/types';
import { TRANSACTION_RECEIPT_FAILED } from '../types';

import closeRed from '../../assets/img/close-red.svg';

const UserErrorComponent = ({
  title,
  message,
  handleCloseClick,
  strings,
  visible,
}) => {
  if (!visible) {
    return <></>;
  }

  const formatMessage = () => {
    if (message.toLowerCase().startsWith('user rejected transaction')) {
      return strings.user_rejected_transaction;
    }
    switch (message) {
      case ERROR_SAME_VALUE:
        return strings.same_value;
      case ERROR_RESOLVE_NAME:
        return strings.resolve_not_set;
      case ADDRESS_ENCODING_ERROR:
        return strings.could_not_encode_address;
      case TRANSACTION_RECEIPT_FAILED:
        return strings.transaction_receipt_failed;
      default:
        return message;
    }
  };

  return (
    <div className="error">
      <button type="button" className="close" onClick={handleCloseClick}>
        <img src={closeRed} alt={strings.close} />
      </button>
      <p>
        <strong>{title}</strong>
      </p>
      <p>{formatMessage()}</p>
    </div>
  );
};

UserErrorComponent.defaultProps = {
  title: 'Error',
  message: 'Error Message',
  handleCloseClick: () => {},
  visible: true,
};

UserErrorComponent.propTypes = {
  strings: propTypes.shape({
    close: propTypes.string.isRequired,
    same_value: propTypes.string.isRequired,
    resolve_not_set: propTypes.string.isRequired,
    could_not_encode_address: propTypes.string.isRequired,
    transaction_receipt_failed: propTypes.string.isRequired,
    user_rejected_transaction: propTypes.string.isRequired,
  }).isRequired,
  title: propTypes.string,
  message: propTypes.string,
  handleCloseClick: propTypes.func,
  visible: propTypes.bool,
};

export default multilanguage(UserErrorComponent);
