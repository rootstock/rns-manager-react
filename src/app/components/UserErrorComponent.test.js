import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import UserErrorComponent from './UserErrorComponent';
import mockStore from '../../../tests/config/mockStore';
import en from '../../languages/en.json';

const store = mockStore({
  close: en.close,
  same_value: en.same_value,
  resolve_not_set: en.resolve_not_set,
  could_not_encode_address: en.could_not_encode_address,
  transaction_receipt_failed: en.transaction_receipt_failed,
  user_rejected_transaction: en.user_rejected_transaction,
});

describe('UserErrorComponent', () => {
  it('renders and matches snapshot', () => {
    const component = mount(<Provider store={store}><UserErrorComponent /></Provider>);
    expect(component).toMatchSnapshot();
  });

  it('renders correct title and message text', () => {
    const component = mount(
      <Provider store={store}>
        <UserErrorComponent title="Test Title" message="Test Message!" />
      </Provider>,
    );

    expect(component.find('strong').text()).toEqual('Test Title');
    expect(component.find('p').at(1).text()).toEqual('Test Message!');
  });

  it('returns blank when visible is false', () => {
    const component = mount(
      <Provider store={store}>
        <UserErrorComponent title="Test Title" message="Test Message!" visible={false} />
      </Provider>,
    );

    expect(component.html()).toEqual('');
  });

  it('returns standard message when sent a constant for no resolver name', () => {
    const component = mount(
      <Provider store={store}>
        <UserErrorComponent message="ERROR_RESOLVE_NAME" visible />
      </Provider>,
    );

    expect(component.find('p').at(1).text()).toEqual(en.resolve_not_set);
  });

  it('returns standard message when sent a constant for same value', () => {
    const component = mount(
      <Provider store={store}>
        <UserErrorComponent message="ERROR_SAME_VALUE" visible />
      </Provider>,
    );

    expect(component.find('p').at(1).text()).toEqual(en.same_value);
  });

  it('Returns user friendly message when the transaction is rejected', () => {
    const component = mount(
      <Provider store={store}>
        <UserErrorComponent message="user rejected transaction bla bla bla" visible />
      </Provider>,
    );

    expect(component.find('p').at(1).text()).toEqual(en.user_rejected_transaction);
  });
});
