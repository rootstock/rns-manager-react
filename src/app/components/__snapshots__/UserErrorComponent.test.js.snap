// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserErrorComponent renders and matches snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(UserErrorComponent)>
    <UserErrorComponent
      currentLanguageCode="en"
      dispatch={[Function]}
      handleCloseClick={[Function]}
      message="Error Message"
      strings={
        Object {
          "close": "Close",
          "could_not_encode_address": "Could not encode address for the network.",
          "resolve_not_set": "Could not resolve domain",
          "same_value": "Input is the same as the value",
          "transaction_receipt_failed": "Transaction failed because it did not complete. This is usually caused by not enough gas being sent.",
        }
      }
      title="Error"
      visible={true}
    >
      <div
        className="error"
      >
        <button
          className="close"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="Close"
            src="mock-image"
          />
        </button>
        <p>
          <strong>
            Error
          </strong>
        </p>
        <p>
          Error Message
        </p>
      </div>
    </UserErrorComponent>
  </Connect(UserErrorComponent)>
</Provider>
`;
